Dockerfiles for TensorRT, CUDA, build from source

vinitra opened the pull request:
- Restructuring dockerfile README
- Adding 3 dockerfiles used for AzureML BYOC scenario

jywu-msft commented on dockerfiles/Dockerfile.tensorrt:
```diff
@@ -0,0 +1,57 @@
+# Dependencies: Ubuntu-core
+RUN apt-get update && \
+  apt-get install -y --no-install-recommends sudo \
+    libopenblas-base \
```
do we really need openblas?

vinitra replied:
removed!

jywu-msft commented on dockerfiles/Dockerfile.cuda:
```diff
@@ -0,0 +1,59 @@
+RUN apt-get update && \
+  apt-get install -y --no-install-recommends sudo \
+    libopenblas-base \
+    python3 \
```
do we need libopenblas, python3 and python3-pip ?

vinitra replied:
removed!

jywu-msft commented on dockerfiles/Dockerfile.tensorrt:
```diff
@@ -0,0 +1,57 @@
+
+# Start the TensorRT build
+RUN /bin/sh ./build.sh --cuda_home /usr/local/cuda \
+--cudnn_home /usr/lib/x86_64-linux-gnu/ --use_tensorrt --tensorrt_home /workspace/tensorrt \
```
I think --tensorrt_home /workspace/tensorrt was a mistake. (though it might not matter)
the tensorrt libs are actually in /usr/lib/x86_64-linux-gnu/

vinitra replied:
The base nVidia TensorRT container has a workspace/tensorrt link with the libs in it!

jywu-msft commented on dockerfiles/Dockerfile.source:
```diff
@@ -0,0 +1,54 @@
+RUN git clone --recursive http://github.com/Microsoft/onnxruntime -b master --single-branch
+WORKDIR /code/onnxruntime
+
+# Start the TensorRT build
```
comment mentions TensorRT ? leftover copy/paste ?

vinitra replied:
Resolved. Thanks for the catch!

jywu-msft commented on dockerfiles/Dockerfile.tensorrt:
```diff
@@ -0,0 +1,57 @@
+# Dockerfile to run ONNXRuntime with TensorRT integration
+
+# nVidia TensorRT Base Image
+FROM nvcr.io/nvidia/tensorrt:19.03-py3
```
good comment. 
we might see if we can directly leverage the same scripts from the source.
i.e. after the git clone of onnxruntime repo, just call install_ubuntu.sh and install_deps.sh  (assuming they aren't installing many more packages than we need)

vinitra replied:
Thanks for the recommendation @HectorSVC!

vinitra replied:
It looks like the [install_ubuntu](https://github.com/Microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/docker/scripts/install_ubuntu.sh) and [install_deps](https://github.com/Microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/docker/scripts/install_deps.sh) scripts that Hector referred to have a lot more dependencies than we need. However, the suggestion to combine common steps in a separate script is a good one.

jywu-msft commented on dockerfiles/Dockerfile.tensorrt:
```diff
@@ -0,0 +1,57 @@
+WORKDIR /work/deps
+
+# Dependencies: Ubuntu-core
+RUN apt-get update && \
```
agreed. we should have separate layer where we think we might change in the future.
and merge RUNs where it makes sense. example could be to do the build and pip install in the same RUN step. see nGraph Dockerfile as example:
https://github.com/otcshare/onnxruntime/blob/master/dockerfiles/Dockerfile.ngraph

jywu-msft commented on dockerfiles/Dockerfile.source:
```diff
@@ -0,0 +1,54 @@
+ENV PATH /code/cmake-3.14.3-Linux-x86_64/bin:${PATH}
+
+# Prepare onnxruntime repository
+RUN git clone --recursive http://github.com/Microsoft/onnxruntime -b master --single-branch
```
i think we'll want an ARG which says which TAG or commitID to build from.  It can default to master.
But we want the ability to build from a release tag or specific commit. (Same comment applies to all our Dockerfiles)

vinitra replied:
Applied from PR #1222

jywu-msft commented on dockerfiles/install_common_deps.sh:
```diff
@@ -0,0 +1,26 @@
+find / -type d -name __pycache__ -prune -exec rm -rf {};
+
+# Dependencies: cmake
+sudo wget --quiet https://github.com/Kitware/CMake/releases/download/v3.14.3/cmake-3.14.3-Linux-x86_64.tar.gz
```
shouldn't need sudo for wget.

vinitra replied:
Removed

jywu-msft commented on dockerfiles/install_common_deps.sh:
```diff
@@ -0,0 +1,26 @@
+conda install -y python=3.6 numpy
+conda clean -aqy
+rm -rf /opt/miniconda/pkgs
+find / -type d -name __pycache__ -prune -exec rm -rf {};
```
why is this find/prune command needed?

vinitra replied:
It was originally included as a requirement from the Azure folks. It seems like it's clearing the conda installation from the python cache (for size reduction?). Any insight @manashgoswami?

vinitra replied:
Removed...

jywu-msft commented on dockerfiles/Dockerfile.source:
```diff
@@ -0,0 +1,22 @@
+ENV PATH /opt/miniconda/bin:/code/cmake-3.14.3-Linux-x86_64/bin:${PATH}
+
+# Dependencies: Ubuntu core, conda, cmake
+COPY install_common_deps.sh /code/install_common_deps.sh
```
it would be better to call the script that's kept in the repo. otherwise, where does the script is frozen and only exists in the image?

vinitra replied:
Good point! Fixed.

jywu-msft commented on dockerfiles/Dockerfile.tensorrt:
```diff
@@ -0,0 +1,28 @@
+# Dockerfile to run ONNXRuntime with TensorRT integration
+
+# nVidia TensorRT Base Image
+FROM nvcr.io/nvidia/tensorrt:19.03-py3
```
replace with FROM nvcr.io/nvidia/tensorrt:19.02-py3

TensorRT 19.03 image is based on TensorRT 5.1
I think it would be better to use TensorRT image 19.02 because ORT source is still using the TRT 5.0 branch of onnx-tensorrt.

vinitra replied:
Updated.

jywu-msft commented on dockerfiles/README.md:
```diff
@@ -18,24 +62,22 @@
-## ONNX Runtime Server (Preview)
-#### Linux 16.04
+## TensorRT Version (Preview)
+#### Linux 16.04, TensorRT 5.1.2
```
let's move to TensorRT 5.0.2
we will update to 5.1 once we can move ORT codebase to 5.1 branch of onnx-tensorrt

vinitra replied:
Updated.

vinitra commented:
Hi @jywu-msft! Addressed your comments in the latest update. Should we include the onnx package as part of the docker image? As of now it only contains the onnxruntime build.

jywu-msft commented:
> Hi @jywu-msft! Addressed your comments in the latest update. Should we include the onnx package as part of the docker image? As of now it only contains the onnxruntime build.

shouldn't need onnx package.
